File: app/src/main/java/me/liwenkun/demo/DemoBook.java
package me.liwenkun.demo;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DemoBook {

    public static void load() {
        try {
            DemoRegister.init();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static final Category ROOT_CATEGORY = new Category("Demo 手册");

    public static class Item implements Comparable<Item> {
        private Category parent;
        protected String name;

        @Override
        public int compareTo(Item o) {
            return name.compareTo(o.name);
        }

        public void setParent(Category parent) {
            this.parent = parent;
        }

        public Category getParent() {
            return parent;
        }
    }

    public static class Category extends Item {

        private final Map<String, DemoItem> demoItems = new HashMap<>();
        private final Map<String, Category> categoryItems = new HashMap<>();

        public Category(String name) {
            this.name = name;
        }

        public void addSubCategory(Category subItem) {
             categoryItems.put(subItem.name, subItem);
        }

        public void addDemoItem(DemoItem demoItem) {
            demoItems.put(demoItem.name, demoItem);
        }

        public List<DemoItem> getDemoItems() {
            List<DemoItem> sortedDemoItems = new ArrayList<>(demoItems.values());
            Collections.sort(sortedDemoItems);
            return sortedDemoItems;
        }

        public List<Category> getSubCategories() {
            List<Category> sortedSubCategories = new ArrayList<>(categoryItems.values());
            Collections.sort(sortedSubCategories);
            return sortedSubCategories;
        }
    }

    public static class DemoItem extends Item {

        Class<?> demoPage;

        public DemoItem(String name, Class<?> demoPage) {
            this.name = name;
            this.demoPage = demoPage;
        }
    }

    public static void add(String demoPath, String title, Class<?> demoActivityClass) {
        String[] pathFragments = demoPath.split("/");
        Category currCategory = ROOT_CATEGORY;
        for (String pathFragment : pathFragments) {
            if ("".equals(pathFragment)) {
                continue;
            }
            Category subCategory = currCategory.categoryItems.get(pathFragment);
            if (subCategory == null) {
                subCategory = new Category(pathFragment);
                subCategory.setParent(currCategory);
                currCategory.addSubCategory(subCategory);
            }
            currCategory = subCategory;
        }
        currCategory.addDemoItem(new DemoItem(title, demoActivityClass));
    }

    public static Category getRootCategory() {
        return ROOT_CATEGORY;
    }
}

File: app/src/main/java/me/liwenkun/demo/demoframework/DemoBook.kt
package me.liwenkun.demo.demoframework

import androidx.lifecycle.MutableLiveData
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.TypeConverters
import me.liwenkun.demo.App.Companion.get
import me.liwenkun.demo.DemoRegister
import me.liwenkun.demo.R

enum class DemoBook {
    INSTANCE;

    private fun load() {
        try {
            Thread {
                DemoRegister.init()
                rootCategory.postValue(ROOT_CATEGORY)
            }.start()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    val rootCategory = MutableLiveData<Category>()

    init {
        load()
    }

    open class Item : Comparable<Item> {
        @Ignore
        var parent: Category? = null

        @ColumnInfo(name = "name")
        lateinit var name: String

        @ColumnInfo(name = "path")
        lateinit var path: String

        constructor()

        constructor(name: String, path: String) {
            this.name = name
            this.path = path
        }

        override fun compareTo(other: Item): Int {
            return name.compareTo(other.name)
        }
    }

    class Category(name: String, path: String) : Item(name, path) {
        private val _demoItems: MutableMap<String, DemoItem> = HashMap()

        internal val _categoryItems: MutableMap<String, Category> = HashMap()

        fun addSubCategory(subItem: Category) {
            _categoryItems[subItem.name] = subItem
        }

        fun addDemoItem(demoItem: DemoItem) {
            _demoItems[demoItem.name] = demoItem
        }

        val demoItems: List<DemoItem>
            get() {
                return _demoItems.values.sorted()
            }

        val categories: List<Category>
            get() {
                return _categoryItems.values.sorted()
            }

        fun clear() {
            _demoItems.clear()
            _categoryItems.clear()
        }
    }

    @Entity(tableName = "demo_items", primaryKeys = ["path"])
    @TypeConverters(
        DemoItemConverters::class
    )
    class DemoItem : Item {
        @ColumnInfo(name = "is_starred")
        var isStarred = false

        @ColumnInfo(name = "demo_page_name")
        lateinit var demoPage: Class<*>

        constructor() : super()

        @Ignore
        constructor(name: String, path: String, demoPage: Class<*>) : super(name, path) {
            isStarred = false
            this.demoPage = demoPage
        }
    }

    companion object {
        private val ROOT_CATEGORY = Category(get().getString(R.string.demo_book), "/")
        @JvmStatic
        fun add(demoPath: String, title: String, demoActivityClass: Class<*>) {
            val pathFragments = demoPath.split('/')
            var currCategory = ROOT_CATEGORY
            val currentPath = StringBuilder()
            for (pathFragment in pathFragments) {
                if ("" == pathFragment) {
                    continue
                }
                currentPath.append('/')
                currentPath.append(pathFragment)
                var subCategory = currCategory._categoryItems[pathFragment]
                if (subCategory == null) {
                    subCategory = Category(pathFragment, currentPath.toString())
                    subCategory.parent = currCategory
                    currCategory.addSubCategory(subCategory)
                }
                currCategory = subCategory
            }
            currentPath.append('/')
            currentPath.append(title)
            val demoItem = DemoItem(title, currentPath.toString(), demoActivityClass)
            get().appDatabase.demoItemDao().insert(demoItem)
            currCategory.addDemoItem(demoItem)
        }
    }
}